azext_iot/iothub/providers/state.py

```diff
@@ -0,0 +1,708 @@
+        if HubAspects.Arm.value in hub_aspects and self.login:
+            raise MutuallyExclusiveArgumentError(constants.LOGIN_WITH_ARM_ERROR)
+        if HubAspects.Arm.value not in hub_aspects and not self.target:
+            raise ResourceNotFoundError(constants.TARGET_HUB_NOT_FOUND_MSG.format(self.hub_name))
+
+        self.delete_aspects(replace, hub_aspects)
+
+        try:
+            with open(state_file, 'r', encoding='utf-8') as f:
+                hub_state = json.load(f)
+
+            self.upload_hub_from_dict(hub_state, hub_aspects)
+            logger.info(constants.UPLOAD_STATE_MSG.format(state_file, self.hub_name))
+
+        except FileNotFoundError:
+            raise FileOperationError(constants.FILE_NOT_FOUND_ERROR.format(state_file))
+
+    def migrate_state(
+        self,
+        orig_hub: Optional[str] = None,
+        orig_rg: Optional[str] = None,
+        orig_hub_login: Optional[str] = None,
+        replace: bool = False,
+        hub_aspects: List[str] = None
+    ):
+        '''Migrates state from original hub to destination hub.'''
+        orig_hub_target = self.discovery.get_target(
+            resource_name=orig_hub,
+            resource_group_name=orig_rg,
+            login=orig_hub_login,
+            auth_type=self.auth_type
+        )
+
+        if not hub_aspects:
+            hub_aspects = HubAspects.list()
+        if HubAspects.Arm.value in hub_aspects and self.login:
+            raise MutuallyExclusiveArgumentError(constants.LOGIN_WITH_ARM_ERROR)
+        if HubAspects.Arm.value not in hub_aspects and not self.target:
+            raise ResourceNotFoundError(constants.TARGET_HUB_NOT_FOUND_MSG.format(self.hub_name))
+
+        # Command modifies hub_aspect - make copy so we can reuse for upload
+        hub_state = self.process_hub_to_dict(orig_hub_target, hub_aspects[:])
+        self.delete_aspects(replace, hub_aspects)
+        self.upload_hub_from_dict(hub_state, hub_aspects)
+
+        logger.info(constants.MIGRATE_STATE_MSG.format(orig_hub, self.hub_name))
+
+    def delete_aspects(self, replace, hub_aspects: List[str] = None):
+        """
+        Delete all necessary hub aspects if the hub exists.
+
+        Delete dataplane aspects only present in hub_aspects.
+        """
+        if self.target and replace:
+            if HubAspects.Configurations.value in hub_aspects:
+                self.delete_all_configs()
+            if HubAspects.Devices.value in hub_aspects:
+                self.delete_all_devices()
+            if HubAspects.Arm.value in hub_aspects:
+                cert_client = self._get_client().certificates
+                # serialize strips name and etag - use as_dict instead
+                certificates = cert_client.list_by_iot_hub(self.rg, self.hub_name).as_dict()
+
+                for cert in tqdm(certificates["value"], desc=constants.DELETE_CERT_DESC, ascii=" #"):
+                    cert_client.delete(self.rg, self.hub_name, cert["name"], cert["etag"])
```

Comment:
The arm template will handle those for us - as in whatever endpoints, routes, ip filters were there before will be overwritten by what is in the arm template.

However, certificates are slightly different from the iot hub arm.

certificates have to be deleted because of possible conflicts. Example scenario:
1. hub has certificate Cert1
2. tries to migrate/import state that has Cert1 with hub_aspects including arm
3. ARM deployment fails because there is a conflict